server/cherrypicker.go

```diff
@@ -0,0 +1,571 @@
+func NewCherryPickerWithOptions(opts CPOptions) *CherryPicker {
+	if opts.Remote == "" {
+		opts.Remote = defaultCherryPickerOpts.Remote
+	}
+	if opts.RepoPath == "" {
+		opts.RepoPath = defaultCherryPickerOpts.RepoPath
+	}
+	return &CherryPicker{
+		options: opts,
+		state:   CPState{},
+		impl:    &defaultCPImplementation{},
+	}
+}
+
+// SetGitHubClient sets a previously configured github client to interact with GH
+func (cp *CherryPicker) SetGitHubClient(ghclient *github.Client) {
+	cp.state.github = ghclient
+}
+
+type CPOptions struct {
+	RepoPath  string // Local path to the repository
+	RepoOwner string // Org of the repo we are using
+	RepoName  string // Name of the repository
+	ForkOwner string
+	Remote    string
+}
+
+var defaultCherryPickerOpts = CPOptions{
+	RepoPath:  ".",
+	Remote:    "origin",
+	ForkOwner: "",
+}
+
+type CPState struct {
+	Repository *git.Repository // Repo object to
+	github     *github.Client  // go-github client
+}
+
+// Actual implementation of the CP interfaces
+type cherryPickerImplementation interface {
+	initialize(context.Context, *CPState, *CPOptions) error
+	readPRcommits(context.Context, *CPState, *CPOptions, *model.PullRequest) ([]*github.RepositoryCommit, error)
+	createBranch(*CPState, *CPOptions, string, *model.PullRequest) (string, error)
+	cherrypickCommits(*CPState, *CPOptions, string, []string) error
+	cherrypickMergeCommit(*CPState, *CPOptions, string, []string, int) error
+	getPRMergeMode(context.Context, *CPState, *CPOptions, *model.PullRequest, []*github.RepositoryCommit) (string, error)
+	findCommitPatchTree(context.Context, *CPState, *CPOptions, *model.PullRequest, []*github.RepositoryCommit) (int, error)
+	GetRebaseCommits(context.Context, *CPState, *CPOptions, *model.PullRequest, []*github.RepositoryCommit) ([]string, error)
+	getPullRequest(context.Context, *CPState, *CPOptions, int) (*model.PullRequest, error)
+	createPullRequest(context.Context, *CPState, *CPOptions, *model.PullRequest, string, string) (*github.PullRequest, error)
+	pushFeatureBranch(*CPState, *CPOptions, string) error
+}
+
+// Initialize checks the environment and populates the state
+func (impl *defaultCPImplementation) initialize(ctx context.Context, state *CPState, opts *CPOptions) error {
+	if state.github == nil {
+		if token := os.Getenv("GITHUB_TOKEN"); token == "" {
+			return errors.New("unable to cherry-pick, github token not found")
+		}
+
+		state.github = github.NewClient(oauth2.NewClient(ctx, oauth2.StaticTokenSource(
+			&oauth2.Token{AccessToken: os.Getenv("GITHUB_TOKEN")},
+		)))
+	}
+
+	// Check the repository path exists
+	if util.Exists(filepath.Join(opts.RepoPath, rebaseMagic)) {
+		return errors.New("there is a rebase in progress, unable to cherry pick at this time")
+	}
+
+	// Open the repo
+	repo, err := git.PlainOpen(opts.RepoPath)
+	if err != nil {
+		return errors.Wrapf(err, "opening repository from %s", opts.RepoPath)
+	}
+
+	// And add it to the state
+	state.Repository = repo
```

Comment:
Is this used anywhere? `go-git` is a big library and I couldn't understand why we would need this?